backend/internal/manager/converge_agent.go

```diff
@@ -0,0 +1,242 @@
+package manager
+
+import (
+	"context"
+	"time"
+
+	ngrok "golang.ngrok.com/ngrok/v2"
+
+	"github.com/ngrok/ngrok-docker-extension/internal/store"
+)
+
+func (m *manager) convergeAgent(ctx context.Context, config store.AgentConfig) error {
+	if config.ExpectedState == AgentStateOffline {
+		if err := m.handleAgentOfflineState(); err != nil {
+			return err
+		}
+	}
+	if config.ExpectedState == AgentStateOnline {
+		if err := m.handleAgentOnlineState(ctx, config); err != nil {
+			return err
+		}
+	}
+	m.agentConfig = config
+	return nil
+}
+
+// handleAgentOfflineState manages disconnecting the agent and setting offline status
+func (m *manager) handleAgentOfflineState() error {
+	// disconnect the agent and set the state to offline
+	m.disconnectAgent()
+	m.setAgentOffline(nil)
+	return nil
+}
```

Comment:
nit: If this never returns anything other than `nil`, then we can drop the `error` from the func signature
```suggestion
func (m *manager) handleAgentOfflineState() {
	// disconnect the agent and set the state to offline
	m.disconnectAgent()
	m.setAgentOffline(nil)
}
```